### PR TITLE
Add RefactorTriggerReason

### PR DESCRIPTION
**Pull request by jessetrinity**
Adds `RefactorTriggerReason` proposed in #35096. Partially covers cases for refactor discoverability #37895.

 This change allows the user to send the code action command (`ctrl + .` in VS and VSCode) to request available refactors at a cursor location (empty span) rather than having to select the entire span of the code to be refactored. The intent is to make refactors more discoverable by allowing a user to "ask" if a refactor is available, rather than having to know in advance that a refactor is applicable to a particular span.

In the future we may want to offer more refactors when the code action command is sent while _non-empty_ span is selected. This change enables that behavior but intentionally does not implement it.

### CODE REVIEW

File: src/server/protocol.ts [hunk 1]
```diff
@@ -546,7 +546,15 @@ namespace ts.server.protocol {
         command: CommandTypes.GetApplicableRefactors;
         arguments: GetApplicableRefactorsRequestArgs;
     }
-    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs;
+    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs & {
+        triggerReason?: RefactorTriggerReason;
+    };
+
+    export type RefactorTriggerReason = RefactorInvokedReason;
```
Comment (amcasey) [inline src/server/protocol.ts:553]: Why re-export it under an additional name?

File: src/server/protocol.ts [hunk 2]
```diff
@@ -546,7 +546,15 @@ namespace ts.server.protocol {
         command: CommandTypes.GetApplicableRefactors;
         arguments: GetApplicableRefactorsRequestArgs;
     }
-    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs;
+    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs & {
+        triggerReason?: RefactorTriggerReason;
```
Comment (amcasey) [inline src/server/protocol.ts:550]: The simplest version of this is a boolean, `isUserInvoked`.  I can see why we might want to be able to generalize that in the future though, so moving to an enum, `{ Invoked, Speculative }`, seems reasonable - it gives us the option of addressing more nuances in the future.  I'm not sure why we'd go more abstract than that though - it seems unlikely we'll need both an enum and an additional payload.

File: src/server/protocol.ts [hunk 3]
```diff
@@ -546,7 +546,15 @@ namespace ts.server.protocol {
...
     }
-    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs;
+    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs & {
+        triggerReason?: RefactorTriggerReason;
+    };
+
+    export type RefactorTriggerReason = RefactorInvokedReason;
+
+    export interface RefactorInvokedReason {
+        kind: "invoked";
```
Comment (amcasey) [inline src/server/protocol.ts:556]: Seems like there should be more than one possible kind?

Comment (amcasey) [reply src/server/protocol.ts:556]: Or maybe we're treating `undefined` as a default kind that means "implicitly invoked"?

File: src/services/refactors/addOrRemoveBracesToArrowFunction.ts [hunk 4]
```diff
@@ -38,7 +38,7 @@ namespace ts.refactor.addOrRemoveBracesToArrowFunction {
 
     function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
         const { file, startPosition } = context;
-        const info = getConvertibleArrowFunctionAtPosition(file, startPosition);
+        const info = getConvertibleArrowFunctionAtPosition(file, startPosition, /*triggerReason*/ { kind: "invoked" });
```
Comment (amcasey) [inline src/services/refactors/addOrRemoveBracesToArrowFunction.ts:41]: This feels very awkward because it's not a request with a trigger reason.  Personally, I would probably change the parameter to a boolean, `isAllowedInFunctionBody`.

File: src/services/refactors/convertExport.ts [hunk 5]
```diff
@@ -31,7 +31,8 @@ namespace ts.refactor {
         const { file } = context;
         const span = getRefactorContextSpan(context);
         const token = getTokenAtPosition(file, span.start);
-        const exportNode = getParentNodeInSpan(token, file, span);
+        // If the span is entirely contained in an export node, check that node.
```
Comment (amcasey) [inline src/services/refactors/convertExport.ts:34]: This seems unrelated?

File: src/services/refactors/extractSymbol.ts [hunk 6]
```diff
@@ -87,7 +87,7 @@ namespace ts.refactor.extractSymbol {
 
     /* Exported for tests */
     export function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
-        const rangeToExtract = getRangeToExtract(context.file, getRefactorContextSpan(context));
+        const rangeToExtract = getRangeToExtract(context.file, getRefactorContextSpan(context), /* triggerReason*/ { kind: "invoked" });
```
Comment (amcasey) [inline src/services/refactors/extractSymbol.ts:90]: As above, faking up a trigger reason here feels awkward.  If we keep it, at least declare a constant.

File: tests/cases/fourslash/refactorConvertImport_partialSelection.ts [hunk 7]
```diff
@@ -3,4 +3,4 @@
 ////import /*a*/d/*b*/, * as n from "m";
 
 goTo.select("a", "b");
-verify.not.refactorAvailable("Convert import");
+verify.refactorAvailable("Convert import");
```
Comment (amcasey) [inline tests/cases/fourslash/refactorConvertImport_partialSelection.ts:6]: I assume the final version of this PR will have additional test coverage.

File: src/server/protocol.ts (hunk 3 above)
Comment (jessetrinity) [reply src/server/protocol.ts:556]: >Or maybe we're treating undefined as a default kind that means "implicitly invoked"?

Yes, that was the idea. It probably makes more sense to have take an explicit `implicit`  kind and treat it the same as an `undefined` `RefactorTriggerReason`.

```ts
export enum RefactorTriggerReason {
    Implicit = "implicit",
    Invoked = "invoked",
}
```

File: src/services/refactors/addOrRemoveBracesToArrowFunction.ts (hunk 4 above)
Comment (jessetrinity) [reply src/services/refactors/addOrRemoveBracesToArrowFunction.ts:41]: I agree it does feels awkward. Maybe it makes more sense to have this be `isImplicit` and default to false? That way we can just ignore the parameter when the function is called for `getEditsForAction`.

File: src/server/protocol.ts (hunk 1 above)
Comment (jessetrinity) [reply src/server/protocol.ts:553]: I think this was from following the pattern for `SignatureHelpTriggerReason`. I guess it makes less sense here.

File: src/server/protocol.ts (hunk 2 above)
Comment (jessetrinity) [reply src/server/protocol.ts:550]: Switching to an enum 

```ts
export enum RefactorTriggerReason {
    Implicit = "implicit",
    Invoked = "invoked",
}
```
seems fine to me. Curious what you think the value of preemptively adding a `Speculative` option would be.

@mjbvz is there potential that VSCode might want to send info in addition to the `triggerReason`, such as a request for a specific "`refactorKind`"?

File: src/services/refactors/convertExport.ts (hunk 5 above)
Comment (jessetrinity) [reply src/services/refactors/convertExport.ts:34]: This and the changes in `convertImport.ts` are just offer refactors more often. They should probably go in a separate PR.

File: src/server/protocol.ts (hunk 2 above)
Comment (amcasey) [reply src/server/protocol.ts]: My `Speculative` is your `Implicit`.

File: src/services/refactors/addOrRemoveBracesToArrowFunction.ts [hunk 8]
```diff
@@ -70,10 +70,12 @@ namespace ts.refactor.addOrRemoveBracesToArrowFunction {
         return { renameFilename: undefined, renameLocation: undefined, edits };
     }
 
-    function getConvertibleArrowFunctionAtPosition(file: SourceFile, startPosition: number): Info | undefined {
+    function getConvertibleArrowFunctionAtPosition(file: SourceFile, startPosition: number, userRequested = true): Info | undefined {
```
Comment (amcasey) [inline src/services/refactors/addOrRemoveBracesToArrowFunction.ts:73]: I can see the convenience of having this be called `userRequested` in all the different refactorings, but, personally, I think it would be clearer to give it a name that's specific to each function you've added it to.  For example, in this case, the parameter could be called something like `considerFunctionBodies` - that would be more meaningful both within the function and at the call-sites that aren't part of a flow that can be either implicit or user-requested.  As a bonus, it would also address my comment above, that this parameter defaults to explicit, whereas the ones in the protocol default to implicit.

File: src/services/refactors/addOrRemoveBracesToArrowFunction.ts [hunk 9]
```diff
@@ -16,8 +16,8 @@ namespace ts.refactor.addOrRemoveBracesToArrowFunction {
     }
 
     function getAvailableActions(context: RefactorContext): readonly ApplicableRefactorInfo[] {
-        const { file, startPosition } = context;
-        const info = getConvertibleArrowFunctionAtPosition(file, startPosition);
+        const { file, startPosition, triggerReason } = context;
+        const info = getConvertibleArrowFunctionAtPosition(file, startPosition, triggerReason === "invoked");
```
Comment (amcasey) [inline src/services/refactors/addOrRemoveBracesToArrowFunction.ts:20]: Unfortunately, JS doesn't have named arguments.  Instead, when we're passing a boolean, we tend to give the parameter name as a `/*comment*/` before the argument (unless the boolean already has a descriptive name, of course).

Comment (jessetrinity) [reply src/services/refactors/addOrRemoveBracesToArrowFunction.ts:20]: It wasn't my intention to pass a named argument, but to pass whatever `triggerReason === "invoked"` evaluates to (`false` if `triggerReason` is `"implicit"` or `undefined`. Did I do that incorrectly?

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 10]
```diff
@@ -6629,7 +6630,10 @@ declare namespace ts.server.protocol {
         command: CommandTypes.GetApplicableRefactors;
         arguments: GetApplicableRefactorsRequestArgs;
     }
-    type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs;
+    type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs & {
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:6633]: This should really be converted to an interface now
